bootstrap/kubeadm/internal/controllers/kubeadmconfig_controller.go

```diff
@@ -321,12 +323,40 @@ func (r *KubeadmConfigReconciler) refreshBootstrapToken(ctx context.Context, con
 		return ctrl.Result{}, err
 	}
 
-	log.Info("Refreshing token until the infrastructure has a chance to consume it")
-	if err := refreshToken(ctx, remoteClient, token, r.TokenTTL); err != nil {
+	secret, err := getToken(ctx, remoteClient, token)
+	if err != nil {
+		return ctrl.Result{}, errors.Wrapf(err, "failed to get bootstrap token secret in order to refresh it")
+	}
+	log = log.WithValues("secretName", secret.Name)
+
+	secretExpiration := bootstrapsecretutil.GetData(secret, bootstrapapi.BootstrapTokenExpirationKey)
+	if len(secretExpiration) == 0 {
+		log.Info(fmt.Sprintf("Token has no valid value for %s, writing new expiration timestamp", bootstrapapi.BootstrapTokenExpirationKey))
+	} else {
+		expiration, err := time.Parse(time.RFC3339, secretExpiration)
+		if err != nil {
+			return ctrl.Result{}, errors.Wrapf(err, "can't parse expiration time of bootstrap token")
+		}
+
+		if expiration.After(time.Now().UTC().Add(r.TokenTTL * 5 / 6)) {
```

Comment:
Refresh if 1/6-th of the TTL has passed. We can still change this, but since `RequeueAfter` is now consistently `r.TokenTTL / 3` (which I consider a sane value even for intermittent errors), the next regular reconciliation would refresh.

Reply:
Let's add a code comment here that explains this

Since this value needs to be tightly coupled with `tokenCheckRefreshOrRotationInterval`, we might want to either define them in the same place in the code or make them depend on each other so we don't inadvertently change one but not the other in the future.